File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        if ((int)nums.size() == k) return nums;
        unordered_map<int, int> mp;
        for (auto &num : nums) mp[num] += 1;
        auto comp = [&mp](int a, int b) {return mp[a] > mp[b];};
        priority_queue<int, vector<int>, decltype(comp)> pq(comp);
        for (auto &p : mp) {
            pq.push(p.first);
            if ((int)pq.size() > k) pq.pop();
        }
        vector<int> ans(k);
        for (int i = k - 1; i >= 0; --i) {
            ans[i] = pq.top();
            pq.pop();
        }
        return ans;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> mp = new HashMap<>();
        int n = nums.length;
        for (int i = 0; i < n; ++i) {
            mp.put(nums[i], mp.getOrDefault(nums[i], 0) + 1);
        }
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> mp.get(a) - mp.get(b));
        for (int a : mp.keySet()) {
            pq.add(a);
            if (pq.size() > k)
                pq.poll();
        }
        int[] ans = new int[k];
        for (int i = 0; i < k; ++i) {
            ans[i] = pq.poll();
        }
        return ans;
    }
}